### DISCUSSION TITLE
MAPファイルの在り処とPrefetch Abort について

### DISCUSSION THREAD

**Discussion by teambeko** [toppers/users-forum#92] [ET-Robocon]
ETロボコン2022 プライマリクラスに参加する者です。

【使用しているPC・環境】
PC:Windows10
走行体:EV3RT HackEV
開発言語: C++

2点質問させてください。

【質問1】
ETロボコン環境ではMAPファイルはどこに出力されますでしょうか?
プログラムカウンタから実行中の関数を調べる際に必要となります。

【質問2】
ロボットの走行中、5回に1回程度ロボットが停止し、以下のエラーが出力されます。
先程質問されたt-hara01様と同じようなエラーですが、Data Abortではなく、
Prefetch Abortとなってます。原因や回避策はありますでしょうか。

====================EXCEPTION DETECTED====================
Prefetch Abort exception occurs.
Task ID: 27
Fault status: 0x0005 (translation fault (1st level))
Fault address: 0xdeafbeef
Segment information:
text segment physical address: 0xc0620000, size: 1048576
data segment physical address: 0xc074b000, size: 1048576
Registers in segment:
lr = 0xc06241d0, virtual address: 0x000041d0 in text segment
Raw exception frame:
pc = 00000008, cpsr = 60000013, lr = c06241d0, r12 = 00000002
r0 = 00000002, r1 = c0741799, r2 = c0741799, r3 = c03fedcf
nest_count = 0, intpri = 0, rundom = 00000001
==========================================================

質問は以上となります。お忙しいところ恐縮ですがよろしくお願いします。

**Comment by hiroakitakada** [comment]
> 【質問1】
> ETロボコン環境ではMAPファイルはどこに出力されますでしょうか?
> プログラムカウンタから実行中の関数を調べる際に必要となります。

TOPPERSの標準環境では，拡張子が .syms のファイルにマップファイルが作られますが，ETロボコン環境でこれが作られるかわかりません。作られない場合には，nmコマンド（クロス開発環境では，実際のコマンド名は，例えば arm-none-eabi-nm のようになります）で作ることができます。

また，プログラムカウンタから実行中のプログラムを調べる場合には，objdumpコマンドも有用です。objdumpはオプションが沢山あるのですが，例えば -d オプションでバイナリプログラムの逆アセンブル結果を出力できます。

**Reply by hiroakitakada** [reply]
> 【質問2】
> ロボットの走行中、5回に1回程度ロボットが停止し、以下のエラーが出力されます。
> 先程質問されたt-hara01様と同じようなエラーですが、Data Abortではなく、
> Prefetch Abortとなってます。原因や回避策はありますでしょうか。

Data Abort が，不正番地へのデータのリード／ライトによる例外であるのに対して，Prefetch Abort は，不正番地へのフェッチ（プログラムのリード）による例外です。つまり，不正な番地に分岐したということです。

原因としては，（関数ポインタによる関数呼び出しを使っていれば）関数ポインタの内容が壊れていて不正な番地に分岐したことや，lrレジスタやスタック領域を壊していてリターン命令で不正な番地に分岐したことが考えられます。Prefetch Abort が起こった状態では，PCが不正な値になっているので，どのプログラムで発生したかを追求するのは面倒です。lrのvirtual addressが0x000041d0になっているので，この番地を含む関数から呼び出された関数が怪しいと思われます。

**Reply by YixiaoLi** [reply]
以下の回答に、objdumpでプログラムを確認する方法を記述しました。ご参考になりましたら幸いです。
> https://github.com/toppers/users-forum/discussions/91#discussioncomment-3707522

**Reply by teambeko** [reply]
高田先生、YixiaoLi先生

ご教示いただきありがとうございます。
残念ながらETロボコンは地区大会敗退でしたが、これで終わりとはせず、
ご教示いただいた内容を基にエラーの解消を目指します。
